### PR TITLE
Add checkpoint file prefix check

### LINKED ISSUE #9465: Misleading error message when running restore op with path to an .index file
Using V2 checkpoint files, calling `saver.restore()` with path to the `.index` returns the following error: 

    Error loading checkpoint from /home/peci1/tradr-git/tradr-ws/src/tradr-simulation/safe_exploration/scripts/ddpg/models/my-model.index: Not found: Tensor name "ActorFullyConnected1/b" not found in checkpoint files /home/peci1/tradr-git/tradr-ws/src/tradr-simulation/safe_exploration/scripts/ddpg/models/my-model.index

I think it'd be nicer to get an error saying "File ... is not a checkpoint. Please, remove the .index extension". This is roughly what you get if you pass any other file.

### PR DESCRIPTION
**Pull request by tedhtchang**
V2 format refers checkpoint files by their file name prefix whereas V1 format refers checkpoint by actual file. I added a check to verify the V2 checkpoint reference is not a file. Also, the return value for saver.save() has changed for V2 so I corrected the doc string.

### CODE REVIEW

**Review by martinwicke — APPROVED** [review]

**Comment by drpngx** [comment]
Apparently that breaks existing behavior.

@tedhtchang could you check the tests?

**Comment by tedhtchang** [comment]
In a freez_graph_test.py log, for example:
https://source.cloud.google.com/results/invocations/4266b4b2-9313-48dd-a120-8579baf5a4e6/targets/%2F%2Ftensorflow%2Fpython%2Ftools:freeze_graph_test/log
"File "/tmp/botexec/bazel-out/k8-opt/bin/tensorflow/python/tools/freeze_graph_test.runfiles/org_tensorflow/tensorflow/python/tools/freeze_graph_test.py", line 151, in testFreezeGraphV1
    self._testFreezeGraph(saver_pb2.SaverDef.V1)

My code has this logic if (os.path.isfile(save_path) and self._write_version != saver_pb2.SaverDef.V1). I am not sure how testFreezeGraphV1 test triggers this error. Let me take a look freez_graph_test.py first.

**Comment by tedhtchang** [comment]
@drpngx The problem is freez_graph_test.py does not pass [saver_pb2.SaverDef.V1](https://github.com/tensorflow/tensorflow/blob/f99275a6a309699c73e1bbebd89ba9aa32e79aa3/tensorflow/python/tools/freeze_graph_test.py#L151) into freez_graph.py. And in freez_graph.py, it does not explicitly initialize the Saver() to use [write_version=saver_pb2.SaverDef.V1](https://github.com/tensorflow/tensorflow/blob/2ebfeda9bad9d1f867205fa8de3db564953ebc23/tensorflow/python/tools/freeze_graph.py#L103) but still be able to restore a V1 checkpoint file.
Which solution do you prefer?
1. Drop the testFreezeGraphV1 test since V1 will be deprecated.
2. Introduce additional argument (write_version=saver_pb2.SaverDef.V2) to the freez_test.py
3. Drop this PR ;however people will not get the useful message when passing any V2 checkpoint file to the restore function.

**Comment by drpngx** [comment]
Option 2 seems to be the best. Thank you!

**Comment by tedhtchang** [comment]
@drpngx Two other failed [tests](https://source.cloud.google.com/results/invocations/35acfd09-b536-4b71-b689-aed4f35f81cb/targets/%2F%2Ftensorflow%2Fcontrib%2Fslim:evaluation_test/log), testAdditionalHooks and testRestoredModelPerformance, also have same problem that the Saver is initialized with [write_version=saver_pb2.SaverDef.V1](https://github.com/tensorflow/tensorflow/blob/4cb0c13c7779da536cac6c682180c5757611b384/tensorflow/contrib/slim/python/slim/evaluation_test.py#L239). I will open another PR to remove write_version=saver_pb2.SaverDef.V1 because I believe it is also fine with V2.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
@tedhtchang is the other PR fixing a test in this one? We try not to submit PRs that have broken tests, to make it easier to bisect. If that's the case, then please fold the other PR into this one.

**Comment by tedhtchang** [comment]
@drpngx I folded other PRs into this one.

**Comment by rmlarsen** [comment]
@drpngx is this good to go?

**Comment by drpngx** [comment]
What I'm saying is that we shouldn't pass the checkpoint version to the `freeze_graph_with_def_protos` that will cause it to *write* in the `V1` format.

**Comment by drpngx** [comment]
OK on second thoughts actually, this is good.

**Comment by rmlarsen** [comment]
@drpngx thanks!

**Comment by FloFoer** [comment]
How am i supposed to use this with the Tensorflow Object Detection API? 
There i have a "Fine-Tune Checkpoint" which is an option in the configuration file that points to a single file. I use the Inception v2 Checkpoint from [here](https://github.com/tensorflow/models/tree/master/research/slim) - again, a single file. 
Previously (r1.5) the training worked just fine with that. However, using TF from the current master or r1.6 (which includes this change here, other than r1.5) it only produces the ValueError you raise in saver.py at line 1749. 
Is the API simply missing support for this and i should post it there or am i doing it wrong?

**Comment by martinwicke** [comment]
@tedhtchang @drpngx it does look like this may be too restrictive. The inception_v2 checkpoint is a v1 checkpoint, and that should be read automatically without raising an error. 

On other words, we cannot assume that we read only v2 checkpoints just because we want to write only v2 checkpoints. Removing the check in saver.py should be enough to fix this, the other changes are still good (I think)

@av8ramit I think this will require a cherry-pick for 1.6 to fix the backwards incompatible change in behavior.

**Comment by av8ramit** [comment]
@gunan who owns the rc1 including cherrypicks.

**Comment by drpngx** [comment]
Yeah, I didn't realize v1 could also be single files. Sorry. We should simply revert the PR.
